urlddedf61e88_dwendelen_CG2_1_tgz-pJ8-RendProjectJ8/src/Vectormath/TransformationMatrix.java

```diff
@@ -0,0 +1,335 @@
+
+    tmp.invdata[0][0] = 1.0; tmp.invdata[0][1] = 0.0; tmp.invdata[0][2] = 0.0; tmp.invdata[0][3] = -trans.data[0];
+    tmp.invdata[1][0] = 0.0; tmp.invdata[1][1] = 1.0; tmp.invdata[1][2] = 0.0; tmp.invdata[1][3] = -trans.data[1];
+    tmp.invdata[2][0] = 0.0; tmp.invdata[2][1] = 0.0; tmp.invdata[2][2] = 1.0; tmp.invdata[2][3] = -trans.data[2];
+    tmp.invdata[3][0] = 0.0; tmp.invdata[3][1] = 0.0; tmp.invdata[3][2] = 0.0; tmp.invdata[3][3] = 1.0;
+
+    return(tmp);
+  }
+
+  /** Creates new scaling matrix, with given vector as scaling factors for x, y and z
+   * @param scale vector that gives 3 scaling factors for x, y and z
+   * @return scaling matrix (new object)
+   */
+  public static TransformationMatrix scale(Vector3 scale) {
+    TransformationMatrix tmp = new TransformationMatrix();
+
+    tmp.data[0][0] = scale.data[0]; tmp.data[0][1] = 0.0; tmp.data[0][2] = 0.0; tmp.data[0][3] = 0.0;
+    tmp.data[1][0] = 0.0; tmp.data[1][1] = scale.data[1]; tmp.data[1][2] = 0.0; tmp.data[1][3] = 0.0;
+    tmp.data[2][0] = 0.0; tmp.data[2][1] = 0.0; tmp.data[2][2] = scale.data[2]; tmp.data[2][3] = 0.0;
+    tmp.data[3][0] = 0.0; tmp.data[3][1] = 0.0; tmp.data[3][2] = 0.0; tmp.data[3][3] = 1.0;
+
+    tmp.invdata[0][0] = 1.0/scale.data[0]; tmp.invdata[0][1] = 0.0; tmp.invdata[0][2] = 0.0; tmp.invdata[0][3] = 0.0;
+    tmp.invdata[1][0] = 0.0; tmp.invdata[1][1] = 1.0/scale.data[1]; tmp.invdata[1][2] = 0.0; tmp.invdata[1][3] = 0.0;
+    tmp.invdata[2][0] = 0.0; tmp.invdata[2][1] = 0.0; tmp.invdata[2][2] = 1.0/scale.data[2]; tmp.invdata[2][3] = 0.0;
+    tmp.invdata[3][0] = 0.0; tmp.invdata[3][1] = 0.0; tmp.invdata[3][2] = 0.0; tmp.invdata[3][3] = 1.0;
+
+    return(tmp);
+  }
+
+  /** Creates rotation matrix around X-axis
+   * @param theta rotation angle in radians. Positive angle obeys right-hand coordinate rule.
+   * @return rotation matrix (new object)
+   */
+  public static TransformationMatrix rotateX(double theta) {
+    TransformationMatrix tmp = new TransformationMatrix();
+
+    tmp.data[0][0] = 1.0; tmp.data[0][1] = 0.0; tmp.data[0][2] = 0.0; tmp.data[0][3] = 0.0;
+    tmp.data[1][0] = 0.0; tmp.data[1][1] = Math.cos(theta); tmp.data[1][2] = -Math.sin(theta); tmp.data[1][3] = 0.0;
+    tmp.data[2][0] = 0.0; tmp.data[2][1] = Math.sin(theta); tmp.data[2][2] = Math.cos(theta); tmp.data[2][3] = 0.0;
+    tmp.data[3][0] = 0.0; tmp.data[3][1] = 0.0; tmp.data[3][2] = 0.0; tmp.data[3][3] = 1.0;
+
+    tmp.invdata[0][0] = 1.0; tmp.invdata[0][1] = 0.0; tmp.invdata[0][2] = 0.0; tmp.invdata[0][3] = 0.0;
+    tmp.invdata[1][0] = 0.0; tmp.invdata[1][1] = Math.cos(theta); tmp.invdata[1][2] = Math.sin(theta); tmp.invdata[1][3] = 0.0;
+    tmp.invdata[2][0] = 0.0; tmp.invdata[2][1] = -Math.sin(theta); tmp.invdata[2][2] = Math.cos(theta); tmp.invdata[2][3] = 0.0;
+    tmp.invdata[3][0] = 0.0; tmp.invdata[3][1] = 0.0; tmp.invdata[3][2] = 0.0; tmp.invdata[3][3] = 1.0;
+
+    return(tmp);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 82:94, 101:113, 120:132.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.